### ISSUE TITLE
Wrong result when run Retina/test.py

### ISSUE THREAD

**Issue by dungnt35** [deepinsight/insightface#769]
![detector_test](https://user-images.githubusercontent.com/35062849/60575328-8e0ea580-9da5-11e9-8eae-482bf35570a2.jpg)
Hi,
I ran /Retina/test.py and get wrong result ( double rectangle on a face) and cant detector another face.
Please tell me what's happend

**Comment by levotrea** [comment]
Hi @dungnt35,

It's hard to say exactly what's wrong here without a minimum of details on what did exactly to get this result. In particular, what model did you use ? I can't be sure, but from the look of it, this problem of accuracy of the results might come from the use of an under-trained model. I suggest you try launching test.py again, but this time using this [pre-trained model](https://www.dropbox.com/s/53ftnlarhyrpkg2/retinaface-R50.zip?dl=0) (you will have to modify the "detector" line in test.py for this).

Hope this helps.

**Comment by dungnt35** [comment]
> Hi @dungnt35,
> 
> It's hard to say exactly what's wrong here without a minimum of details on what did exactly to get this result. In particular, what model did you use ? I can't be sure, but from the look of it, this problem of accuracy of the results might come from the use of an under-trained model. I suggest you try launching test.py again, but this time using this [pre-trained model](https://www.dropbox.com/s/53ftnlarhyrpkg2/retinaface-R50.zip?dl=0) (you will have to modify the "detector" line in test.py for this).
> 
> Hope this helps.

Thanks for your help.
I rune default test.py with "detector = RetinaFace('./model/R50', 0, gpuid, 'net3')"
I download pretrain model in readme from Retina folder.
Pretrained Model: RetinaFace-R50.
Can you give me your pretrain model for this case.
Thanks so much

**Comment by levotrea** [comment]
Well, I personally used the model from the link I put in my first answer ([](https://www.dropbox.com/s/53ftnlarhyrpkg2/retinaface-R50.zip?dl=0)), which seems to be the one you are using as well.

This might be a stupid question from my part, but you did place the model in Retina/model/ , right ?
(Retina/ being the folder in which test.py is)

Can you maybe give me a link to the image you launched test.py on, so that I can at least try it myself and see if I get the same result as yours ?

**Comment by dungnt35** [comment]
> Well, I personally used the model from the link I put in my first answer (https://www.dropbox.com/s/53ftnlarhyrpkg2/retinaface-R50.zip?dl=0), which seems to be the one you are using as well.
> 
> This might be a stupid question from my part, but you did place the model in Retina/model/ , right ?
> (Retina/ being the folder in which test.py is)
> 
> Can you maybe give me a link to the image you launched test.py on, so that I can at least try it myself and see if I get the same result as yours ?
> 
> **Edit :**
> 
> I launched my test.py on the image you provided in your first message with the pre-trained model I gave you, and it found all of the faces.
> 
> (As I used your image, the bounding boxes from your execution are still there, but my execution produced only one box for each face).
> 
> ![analyzed_github_test](https://user-images.githubusercontent.com/38857756/60589760-8c8bb000-9d9a-11e9-9795-1d6036f512ac.jpg)
> 
> So I'm not sure what your issue is. My only advice for the moment would be to try downloading the model from the link I gave you, and in the "detector" line from test.py, replace `./model/R50` with the absolute path to the newly downloaded model (not forgetting the prefix of the model itself).
> So the detector line would look something like this : `detector = RetinaFace("/xxx/retinaface-R50/R50", 0, gpuid, 'net3')`, where "xxx" is the absolute path to the downloaded model.

Here is my code from the detector line :
detector = RetinaFace('../models/R50', 0, gpuid, 'net3')
And here is my pretrain model :
![wrong](https://user-images.githubusercontent.com/35062849/60591112-ed7dad00-9dc7-11e9-85e8-5d88ff36e14e.PNG)
I ran it again but have same result :(
And another my stupid question :
Why have double, triple rectangle on a face?
Thanks so much.

**Comment by levotrea** [comment]
> Here is my code from the detector line :
> detector = RetinaFace('../models/R50', 0, gpuid, 'net3')
> And here is my pretrain model :
> ![wrong](https://user-images.githubusercontent.com/35062849/60591112-ed7dad00-9dc7-11e9-85e8-5d88ff36e14e.PNG)
> I ran it again but have same result :(
> And another my stupid question :
> Why have double, triple rectangle on a face?
> Thanks so much.

Is it the same model you were using before or is it the one I gave you in the link ?

As for the multiple boxes, I am really not sure exactly how RetinaFace works, but I think it may start by finding many potential windows (or rectangles if you prefer) in which faces might be found in the image, before eliminating most of them after further analysis (in particular merging the boxes that are very close from one another). So in the end, if the analysis was precise enough, there should only be one box left for each face (and each face should have its box).

But this is only me speculating here. What I explained is roughly how an other method (MTCNN) worked, and I think RetinFace might have some similarities with it. But as I am not part of the research team, just take what I said as a possible explanation of why there are multiple boxes in your image, but I may very well be wrong.

Though, the fact that there are multiple boxes would mean that the analysis wasn't precise enough, or maybe not complete (?), and it still makes me think that your issue is coming from the use an inefficient model.

I'm sorry that this is the only answer I have to give you, but if you indeed used the model I gave you in my previous message and still get this inaccurate result, then I don't know where your issue might be coming from. If you didn't replace your model by the one I gave you, then you should try it. I'm not saying this will solve your issue, but it works fine in my case.

Last question : Did you modify any of the code from retinaface.py or test.py (apart from the detector line) ?

**Comment by dungnt35** [comment]
> > Here is my code from the detector line :
> > detector = RetinaFace('../models/R50', 0, gpuid, 'net3')
> > And here is my pretrain model :
> > ![wrong](https://user-images.githubusercontent.com/35062849/60591112-ed7dad00-9dc7-11e9-85e8-5d88ff36e14e.PNG)
> > I ran it again but have same result :(
> > And another my stupid question :
> > Why have double, triple rectangle on a face?
> > Thanks so much.
> 
> Is it the same model you were using before or is it the one I gave you in the link ?
> 
> As for the multiple boxes, I am really not sure exactly how RetinaFace works, but I think it may start by finding many potential windows (or rectangles if you prefer) in which faces might be found in the image, before eliminating most of them after further analysis (in particular merging the boxes that are very close from one another). So in the end, if the analysis was precise enough, there should only be one box left for each face (and each face should have its box).
> 
> But this is only me speculating here. What I explained is roughly how an other method (MTCNN) worked, and I think RetinFace might have some similarities with it. But as I am not part of the research team, just take what I said as a possible explanation of why there are multiple boxes in your image, but I may very well be wrong.
> 
> Though, the fact that there are multiple boxes would mean that the analysis wasn't precise enough, or maybe not complete (?), and it still makes me think that your issue is coming from the use an inefficient model.
> 
> I'm sorry that this is the only answer I have to give you, but if you indeed used the model I gave you in my previous message and still get this inaccurate result, then I don't know where your issue might be coming from. If you didn't replace your model by the one I gave you, then you should try it. I'm not saying this will solve your issue, but it works fine in my case.
> 
> Last question : Did you modify any of the code from retinaface.py or test.py (apart from the detector line) ?

Thanks for your help.
I still have same result.
I think i will change another detector like mtcnn or ssh
Can i have skype id , or face book or anythink to contact with you?

**Comment by levotrea** [comment]
> Thanks for your help.
> I still have same result.
> I think i will change another detector like mtcnn or ssh
> Can i have skype id , or face book or anythink to contact with you?

Well, it's a shame that it doesn't work for you. and I don't see what could influence the accuracy of the results other than the model. Maybe @nttstar knows of other factors that can lead to such inaccurate results ? (The GPU, or CUDA version for example ?)

The simplest way to contact me would be by e-mail : antoine.levotre@grenoble-inp.org (I don't use Skype, and I prefer not to use Facebook for any professional purpose). I'm not sure I can't be of any help though, as I have very limited knowledge in AI. But I would be glad to provide any help that I can.
